IOS Dialog - Extensions

PThorpe92 opened the pull request:
Going to mark this as draft until it can be reviewed. Check happens fine, dialog comes up immediately and REPL cancels out but due to my setup here I am unable to test very well. 

https://github.com/mit-cml/appinventor-sources/issues/2836

AppInventorWorkerBee commented:
Can one of the admins verify this patch?

ewpatton commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1427,15 +1427,13 @@ Blockly.ReplMgr.startRepl = function(already, chromebook, emulator, usb) {
         this.hardreset(this.formName);       // Tell aiStarter to kill off adb
     }
 };
-
```
There's no need to remove this line.

ewpatton commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1427,15 +1427,13 @@ Blockly.ReplMgr.startRepl = function(already, chromebook, emulator, usb) {
     }
     return retval;
 };
-
```
Ditto

ewpatton commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1477,7 +1475,16 @@ Blockly.ReplMgr.getFromRendezvous = function() {
                 rs.hasfetchassets = rs.android || rs.webrtc;
-
+                if (!(rs.android) && Blockly.ReplMgr.hasExtensions()) {
+     var content = "Extentions are not currently supported for IOS Devices";
```
We should put the message in Blockly.Msg so that it can be internationalized. See blocklyeditor/src/msg/ai_blockly/messages.json.

ewpatton commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1691,7 +1707,6 @@ Blockly.ReplMgr.resendAssetsAndExtensions = function() {
         Blockly.ReplMgr.loadExtensions();
     });
 };
-
```
Restore the blank line here.

ewpatton commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1507,6 +1514,15 @@ Blockly.ReplMgr.getFromRendezvous = function() {
 };
 
+Blockly.ReplMgr.hasExtensions = function() {
+   var extensions = top.AssetManager_getExtensions();
```
Make sure the indentation in your changes matches the rest of the file.

ewpatton commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1507,6 +1514,15 @@ Blockly.ReplMgr.getFromRendezvous = function() {
 
+Blockly.ReplMgr.hasExtensions = function() {
+   var extensions = top.AssetManager_getExtensions();
+   if (extensions.length == 0) {
```
You can simplify this to `return extensions.length > 0;`

ewpatton reviewed (changes requested):
A few minor comments to address.

PThorpe92 commented:
My bad I had pushed the wrong branch and deleted the one I intended to submit... didn't check out from rough draft hence random line deletions. Fixed with requested changes.

ewpatton commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1471,6 +1471,15 @@ Blockly.ReplMgr.getFromRendezvous = function() {
                 rs.versionurl = 'http://' + json.ipaddr + ':8001/_getversion';
                 rs.baseurl = 'http://' + json.ipaddr + ':8001/';
                 rs.android = !(new RegExp('^i(pad)?os$').test((json.os || 'Android').toLowerCase()));
+                if (!(rs.android) && Blockly.ReplMgr.HasExtensions()){
```
The call should be `hasExtension`, not `HasExtension`. The latter is undefined.

ewpatton reviewed (changes requested):
Some good progress here. I have identified a few more issues with this update. Please see my inline comments.

PThorpe92 commented:
Really shouldn't have rushed that, my fault

ewpatton commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1471,6 +1471,15 @@ Blockly.ReplMgr.getFromRendezvous = function() {
+                        top.BlocklyPanel_indicateDisconnect();
+                        rs.connection = null;
+                    }); 
+                ios_dialog.display();
```
So, it turns out that as long as we're giving the button name the dialog will appear automatically. By calling this here, we end up with two copies of the dialogs, one of which can't be dismissed.

PThorpe92 replied:
Ok I wasn't sure about that because when I tested it locally it didn't display on it's own, but calling the function displayed two

ewpatton commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1471,6 +1471,14 @@ Blockly.ReplMgr.getFromRendezvous = function() {
+                if (!(rs.android) && Blockly.ReplMgr.hasExtensions()){
+                    var ios_dialog = new Blockly.Util.Dialog(Blockly.Msg.EXTENSIONS, Blockly.Msg.EXTENSIONS_iOS, Blockly.Msg.REPL_CANCEL, true, null, 0, function(){
+                        ios_dialog.hide();
+                        top.ReplState.state = Blockly.ReplMgr.rsState.IDLE;
+                        top.BlocklyPanel_indicateDisconnect();
+                        rs.connection = null;
```
I think we want to move these into the outer scope after the dialog is made and then return. Right now as implemented the system will keep trying to continue the process, and how far it gets will depend on how long it takes for someone to close the dialog. Instead, we probably want to cancel the connection immediately and return as soon as the dialog is shown (the dialog doesn't block).

PThorpe92 replied:
Oh ok makes sense. If the only thing in the callback is dialog.hide() can that be put as a direct argument the the new dialog or do I still have to define the anonymous function with it inside?
e.g ios_dialog = new Blockly.Util.Dialog(Blockly.Msg.EXTENSIONS, Blockly.Msg.EXTENSIONS_iOS, Blockly.Msg.REPL_CANCEL, true, null, 0, ios_dialog.hide()) ? I would assume define a function but I'd rather ask just in case

ewpatton commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1471,6 +1471,14 @@ Blockly.ReplMgr.getFromRendezvous = function() {
+                    var ios_dialog = new Blockly.Util.Dialog(Blockly.Msg.EXTENSIONS, Blockly.Msg.EXTENSIONS_iOS, Blockly.Msg.REPL_CANCEL, true, null, 0, function(){
+                        ios_dialog.hide();
+                    }); 
+                }
```
You still need to `return` here at the end of the `if`, otherwise the rest of the function will keep running and try to establish a connection.

ewpatton replied:
@PThorpe92 Don't worry it happens to all of us at some point.

ewpatton replied:
Hmm. So the progress bar still appears even though the connection process has been stopped. I think we need to move the two calls to set up the progress bar o lines 1466-1467 to after the `if`. Since everything between there and here is setting up variables this should be safe.

ewpatton replied:
Note that for testing you could always intentionally set rs.android = false, but just make sure to take it out before committing your changes.

PThorpe92 replied:
Yeah I actually tested it again instead of making panicky changes one at a time. After pushing the wrong branch initially that just set it off to a terrible start.
So even after moving those calls the progress bar was still appearing in front of the warning dialog still so I added the rs.dialog.hide() and it seems to work fine now

ewpatton commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1462,15 +1462,25 @@ Blockly.ReplMgr.getFromRendezvous = function() {
                 rs.versionurl = 'http://' + json.ipaddr + ':8001/_getversion';
                 rs.baseurl = 'http://' + json.ipaddr + ':8001/';
                 rs.android = !(new RegExp('^i(pad)?os$').test((json.os || 'Android').toLowerCase()));
+                if (!(rs.android) && Blockly.ReplMgr.hasExtensions()){
```
Code style: Add space between closing parens and curly brace.

ewpatton commented on appinventor/blocklyeditor/src/replmgr.js:
```diff
@@ -1462,15 +1462,25 @@ Blockly.ReplMgr.getFromRendezvous = function() {
+                    top.ReplState.state = Blockly.ReplMgr.rsState.IDLE;
+                    top.BlocklyPanel_indicateDisconnect();
+                    rs.connection = null;
+                    var ios_dialog = new Blockly.Util.Dialog(Blockly.Msg.EXTENSIONS, Blockly.Msg.EXTENSIONS_iOS, Blockly.Msg.REPL_CANCEL, true, null, 0, function(){
```
Code style: Add space between closing parens and curly brace.

ewpatton reviewed (changes requested):
Two minor code style things but otherwise LGTM.

ewpatton commented:
@AppInventorWorkerBee ok to test

ewpatton reviewed (approved):
LGTM
